gobblin-service/src/main/java/org/apache/gobblin/service/modules/flowgraph/FlowGraph.java

```diff
@@ -0,0 +1,102 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.flowgraph;
+
+import java.util.Collection;
+
+import org.apache.gobblin.annotation.Alpha;
+
+
+/**
+ * An interface for {@link FlowGraph}. A {@link FlowGraph} consists of {@link DataNode}s and {@link FlowEdge}s.
+ * The interface provides methods for adding and removing {@link DataNode}s and {@link FlowEdge}s to the {@link FlowGraph}.
+ * In addition the interface provides methods to return factory classes for creation of {@link DataNode}s and {@link FlowEdge}s.
+ */
+
+@Alpha
+public interface FlowGraph {
+  public Collection<DataNode> getNodes();
+
+  /**
+   * Add a {@link DataNode} to the {@link FlowGraph}
+   * @param node {@link DataNode} to be added
+   * @return true if {@link DataNode} is added to the {@link FlowGraph} successfully.
+   */
+  public boolean addDataNode(DataNode node);
+
+  /**
+   * Add a {@link FlowEdge} to the {@link FlowGraph}
+   * @param edge {@link FlowEdge} to be added
+   * @return true if {@link FlowEdge} is added to the {@link FlowGraph} successfully.
+   */
+  public boolean addFlowEdge(FlowEdge edge);
+
+  /**
+   * Remove a {@link DataNode} and all its incident edges from the {@link FlowGraph}
+   * @param node {@link DataNode} to be removed
+   * @return true if {@link DataNode} is removed from the {@link FlowGraph} successfully.
+   */
+  public boolean deleteDataNode(DataNode node);
+
+  /**
+   * Remove a {@link DataNode} and all its incident edges from the {@link FlowGraph}
+   * @param nodeId identifier of the {@link DataNode} to be removed
+   * @return true if {@link DataNode} is removed from the {@link FlowGraph} successfully.
+   */
+  public boolean deleteDataNode(String nodeId);
+
+  /**
+   * Remove a {@link FlowEdge} from the {@link FlowGraph}
+   * @param edge to be removed
+   * @return true if edge is removed from the {@link FlowGraph} successfully.
+   */
+  public boolean deleteFlowEdge(FlowEdge edge);
+
+  /**
+   * Remove a {@link FlowEdge} from the {@link FlowGraph}
+   * @param edgeId label of the edge to be removed
+   * @return true if edge is removed from the {@link FlowGraph} successfully.
+   */
+  public boolean deleteFlowEdge(String edgeId);
+
+  /**
+   * Get the factory class for creating {@link DataNode}s to be added to the {@link FlowGraph}.
+   * @return a {@link DataNodeFactory} implementation.
+   */
+  public DataNodeFactory getDataNodeFactory();
+
+  /**
+   * Get the factory class for creating {@link FlowEdge}s to be added to the {@link FlowGraph}.
+   * @return a {@link FlowEdgeFactory} implementation.
+   */
+  public FlowEdgeFactory getFlowEdgeFactory();
```

Comment:
Same as above

Reply:
See comment above.

Reply:
Is it better to decouple the factory from flow graph?  Like how we treat the data node factory, edge factory class is specified by configuration.